cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -0,0 +1,197 @@
+    def protocol(self) -> str:
+        """Protocol used to resolve this dependency.
+
+        It is parsed from the resolution locator.
+        """
+        matches = _PACKAGE_PROTOCOL_REGEX.search(self.locator)
+
+        if matches:
+            return matches[0][1:-1]
+
+        raise Exception("Invalid locator string.")
+
+    @property
+    def name_in_locator(self) -> str:
+        """Name of the package as parsed from its locator string."""
+        parts = self.locator.split("@")
+
+        # aliased package, starts with @
+        if parts[0] == "":
+            return "@" + parts[1]
+
+        return parts[0]
+
+    @classmethod
+    def from_info_string(cls, info: str) -> "Package":
+        """Create a Package from the output of yarn info."""
+        data = json.loads(info)
+
+        return Package(
+            locator=data["value"],
+            version=data["children"]["Version"],
+            dependencies=[dep for dep in data["children"].get("Dependencies", [])],
+        )
+
+
+def resolve_dependency_info(source_dir: RootedPath) -> list[Component]:
+    """Fetch and parse dependency info from the 'yarn info' output.
+
+    This function also performs a validation to ensure that the current
+    yarn project can be processed.
+
+    :raises PackageRejected: if the validation fails.
+    :raises SubprocessCallError: if the 'yarn info' command fails.
+    """
+    cmd = ["info", "-AR", "--json", "--cache"]
+
+    result = run_yarn_cmd(cmd, source_dir)
+
+    packages = [Package.from_info_string(info) for info in result.splitlines()]
+
+    _validate_protocols_in_dependencies(packages)
+    _vet_git_dependencies(packages)
+
+    return [_create_component_from_package(package) for package in packages]
+
+
+def _validate_protocols_in_dependencies(packages: list[Package]) -> None:
+    """Validate that Cachi2 can process all protocols present in the dependencies.
+
+    :raises PackageRejected: if a unprocessable protocol is found.
+    """
+    pass
+
+
+def _vet_git_dependencies(packages: list[Package]) -> None:
+    """Vet a Git dependency if it contains forbidden lifecycle scripts.
+
+    :raises PackageRejected: if a dependency containing a forbidden lifecycle
+    script is found.
+    """
+    pass
+
+
+def _resolve_package_name(package: Package) -> str:
+    """Resolve the package's real name based on its protocol.
+
+    Non-registry deps might have names in their locator string that
+    differ from the one in their package.json file.
+
+    Look at the package.json name for every non-registry dependency.
+    """
+    if package.protocol == "npm":
+        return package.name_in_locator
+
+    if package.protocol in ["file"]:
+        return _resolve_remote_file_dependency_name(package)
+
+    if package.protocol in ["portal", "link", "workspace"]:
+        return _resolve_local_file_dependency_name(package)
+
+    if package.protocol in ["git", "github"]:
+        return _resolve_git_dependency_name(package)
+
```

Comment:
Ah, I see. We don't need to custom-fetch them, we can just call `yarn install` and read the package.jsons in the zip files (`yarn info` tells you where those zip files are)

Reply:
Fixed it. See https://github.com/containerbuildsystem/cachi2/pull/309#discussion_r1318218711.